plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -0,0 +1,321 @@
+					   error)) {
+		g_prefix_error (error, "failed to claim interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+
+static gboolean
+fu_cros_ec_usb_device_probe (FuUsbDevice *device, GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	/* very much like usb_updater2's usb_findit() */
+
+	if (!fu_cros_ec_usb_device_find_interface (device, error)) {
+		g_prefix_error (error, "failed to find update interface: ");
+		return FALSE;
+	}
+
+	if (self->chunk_len == 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "wMaxPacketSize isn't valid: %" G_GUINT16_FORMAT,
+			     self->chunk_len);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+do_xfer (FuCrosEcUsbDevice * cros_ec_device, guint8 *outbuf, gsize outlen,
+	 guint8 *inbuf, gsize inlen, gboolean allow_less,
+	 gsize *rxed_count, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (cros_ec_device));
+	FuCrosEcUsbDevice *self = cros_ec_device;
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer (usb_device, self->ep_num,
+						 outbuf, outlen,
+						 &actual, BULK_SEND_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only sent %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer (usb_device,
+						 self->ep_num | 0x80,
+						 inbuf, inlen,
+						 &actual, BULK_RECV_TIMEOUT_MS,
+						 NULL, error)) {
+			return FALSE;
+		}
+		if (actual != inlen && !allow_less) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only received %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count != NULL)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	gsize rxed_size;
+	guint8 inbuf[64];
+	gsize actual = 0;
+	guint32 error_code;
+	struct update_frame_header ufh;
+
+	/*
+	 * Need to be backwards compatible, communicate with targets running
+	 * different protocol versions.
+	 */
+	union {
+		struct first_response_pdu rpdu;
+		guint32 legacy_resp;
+	} start_resp;
+
+	/* send start request */
+
+	/* flush all data from endpoint to recover in case of error */
+	while (g_usb_device_bulk_transfer (usb_device,
+					   self->ep_num | 0x80,
+					   inbuf, self->chunk_len,
+					   &actual, 10, NULL, NULL))
+		g_debug ("flush!\n");
```

Comment:
Some more conversation around this flushing routine.  Since it's a while loop matching on the success return case, waiting for the fail case to break seems counterintutive.  Can it be a fixed number of calls in a for loop (or a `fu_device_retry`)?  Or does it really need to be bulk transfer until it fails?

If done with `fu_device_retry` you can probably set the exact circumstance that it should be a success and fail, how many retries and it will be much more obvious what's going on.

Reply:
Hi @superm1 

I will look into implementing it as a fu_device_retry.

From what I understand about what this is seeking to accomplish, the normal case would result in a failed bulk read on the 1st attempt, since when the program starts, or when the usb device is attached, there was no command sent to the device, and no data is ready to be sent back to the host (since nothing was requested).

In the case where a flush is necessary, it would actually read bytes out and the bulk transfer would succeed, indicating that something was requested by a previous iteration of this updater (but maybe it crashed or was interrupted leaving the data buffered by the EC).

I think we can set some upper limit, but I'm still mulling over what that number should be.

Reply:
Thanks for explaining it.  I think an upper limit makes sense so you don't have the possibility to hang the daemon in a loop here because of a possible firmware bug.

Reply:
Hi @superm1.

I've gone ahead and implemented the fu_device_retry. After experimenting with purposefully wedging the device, I got the handle on how fu_device_retry works.

I think I might actually refactor another part of the _setup() in order to use a separate fu_device_retry as well. Thanks for the tip.